interoperator/pkg/watches/channel.go

```diff
@@ -0,0 +1,252 @@
+			continue
+		}
+	}
+	go func() {
+		<-stopCh
+		for _, clusterWatcher := range clusterWatcherList {
+			if clusterWatcher != nil {
+				clusterWatcher.stopper <- struct{}{}
+			}
+		}
+		// close events after all clusterWatchers are done
+		waitgroup.Wait()
+		close(events)
+		log.Info("closed watch channel", "controller", controllerName)
+
+	}()
+	log.Info("created watch channel", "controller", controllerName)
+	return events, stopCh, nil
+}
+
+type clusterWatcher struct {
+	controllerName  string
+	clusterID       string
+	dynamicClient   dynamic.Interface
+	discoveryClient discovery.DiscoveryInterface
+	gvkList         []schema.GroupVersionKind
+	gvrList         []schema.GroupVersionResource
+	watchers        []*watcher
+	events          chan<- event.GenericEvent
+	stop            <-chan struct{}
+	// stopper is the write side of the stop channel. They should have the same value.
+	stopper   chan<- struct{}
+	waitgroup *sync.WaitGroup
+}
+
+func (cw *clusterWatcher) computeGVRList() {
+	has := func(verbs []string, verb string) bool {
+		for _, val := range verbs {
+			if val == verb {
+				return true
+			}
+		}
+		return false
+	}
+	gvrList := make([]schema.GroupVersionResource, 0)
+	for _, gvk := range cw.gvkList {
+		apiResourceList, err := cw.discoveryClient.ServerResourcesForGroupVersion(gvk.GroupVersion().String())
+		if err != nil {
+			log.Error(err, "failed to discover resource for watch",
+				"clusterID", cw.clusterID, "gvk", gvk, "controller", cw.controllerName)
+			continue
+		}
+		for _, apiResource := range apiResourceList.APIResources {
+			if gvk.Kind == apiResource.Kind && has(apiResource.Verbs, "watch") {
+				gvrList = append(gvrList, gvk.GroupVersion().WithResource(apiResource.Name))
+			}
+		}
+	}
+	cw.gvrList = gvrList
+}
+
+func (cw *clusterWatcher) start() error {
+	cw.computeGVRList()
+	cw.watchers = make([]*watcher, len(cw.gvrList))
+	for i, gvr := range cw.gvrList {
+		stop := make(chan struct{})
+		cw.watchers[i] = &watcher{
+			controllerName: cw.controllerName,
+			clusterID:      cw.clusterID,
+			dynamicClient:  cw.dynamicClient,
+			gvr:            gvr,
+			events:         cw.events,
+			stop:           stop,
+			stopper:        stop,
+			waitgroup:      cw.waitgroup,
+		}
+		err := cw.watchers[i].start()
+		if err != nil {
+			log.Error(err, "failed to start watch on cluster for resource",
+				"clusterID", cw.clusterID, "gvr", gvr, "controller", cw.controllerName)
+			continue
+		}
+	}
+	go func() {
+		<-cw.stop
+		for _, watcher := range cw.watchers {
+			watcher.stopper <- struct{}{}
+		}
+	}()
+	return nil
+}
+
+type watcher struct {
+	controllerName string
+	clusterID      string
+	dynamicClient  dynamic.Interface
+	gvr            schema.GroupVersionResource
+	events         chan<- event.GenericEvent
+	stop           <-chan struct{}
+	// stopper is the write side of the stop channel. They should have the same value.
+	stopper   chan<- struct{}
```

Comment:
What is this for?

Reply:
`stop `and `stopper `are the same channel. `Watchers `listen on the read only channel `stop `for the stop event. `clusterWatcher `issues stop signal to watcher using the write only `stopper `channel. 

It could be just one field `stop chan struct{}`